File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp

class Solution {
public:
    int n;

    //  reference pass kro  → no copy
    bool isPalindrome(int l, int r, const string &s) {
        while (l < r) {
            if (s[l] != s[r]) return false;
            l++;
            r--;
        }
        return true;
    }

    void solve(vector<vector<string>> &result, vector<string> &temp, const string &s, int idx) {
        if (idx == n) {
            result.push_back(temp);
            return;
        }

        for (int i = idx; i < n; i++) {
            if (isPalindrome(idx, i, s)) {
                temp.push_back(s.substr(idx, i - idx + 1)); // add substring
                solve(result, temp, s, i + 1);                 // recurse
                temp.pop_back();                            // backtrack
            }
        }
    }

    vector<vector<string>> partition(string s) {
        n = s.length();
        vector<vector<string>> result;
        vector<string> temp;
        solve(result, temp, s, 0);
        return result;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:

    def __init__(self):
        self.n = 0

    # Reference pass in C++ → Python passes objects by reference
    def isPalindrome(self, l, r, s):
        while l < r:
            if s[l] != s[r]:
                return False

            l += 1
            r -= 1

        return True

    def solve(self, result, temp, s, idx):

        # Base case
        if idx == self.n:
            result.append(temp.copy())
            return

        for i in range(idx, self.n):

            if self.isPalindrome(idx, i, s):

                # Take
                temp.append(s[idx:i + 1])

                # Explore
                self.solve(result, temp, s, i + 1)

                # Backtrack
                temp.pop()

    def partition(self, s):
        self.n = len(s)

        result = []
        temp = []

        self.solve(result, temp, s, 0)

        return result